protocols/kad/src/behaviour.rs

```diff
@@ -1513,11 +1559,12 @@ where
                 if let Some(query) = self.queries.get_mut(&user_data) {
                     query.on_success(&source, vec![]);
                     if let QueryInfo::PutRecord {
-                        phase: PutRecordPhase::PutRecord { num_results, .. }, quorum, ..
+                        phase: PutRecordPhase::PutRecord { success, .. }, quorum, ..
                     } = &mut query.inner.info {
-                        *num_results += 1;
-                        if *num_results >= quorum.get() {
-                            query.finish()
+                        success.push(source);
+                        if success.len() >= quorum.get() {
+                            let peers = success.clone();
+                            query.may_finish(peers)
```

Comment:
```suggestion
                            query.try_finish(success.iter())
```
Wouldn't that work? Or at least `success.iter().cloned()`?

Reply:
As the iterator is still lazy it keeps borrowing `query`:

```
error[E0499]: cannot borrow `*query` as mutable more than once at a time
    --> protocols/kad/src/behaviour.rs:1561:29
     |
1558 |                     } = &mut query.inner.info {
     |                         --------------------- first mutable borrow occurs here
...
1561 |                             query.may_finish(success.iter().cloned())
     |                             ^^^^^            ------- first borrow later used here
     |                             |
     |                             second mutable borrow occurs here

error: aborting due to previous error
```

I am not aware of an easy way around this.